File: cpp/nextgreaterelementii.cpp
class Solution {
public:
    vector<int> nextGreaterElements(vector<int>& nums) {
        int n = nums.size();
        stack<int> s;
        vector<int> res(n, -1);
        for (int i = 0; i < n*2; i++) {
            int num = nums[i%n];
            while (!s.empty() and num > nums[s.top()]) {
                res[s.top()] = num;
                s.pop();
            }
            if (i < n) s.push(i);
        }
        return res;
    }
};

File: python/nextgreaterelementii.py
class Solution(object):
    def nextGreaterElements(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        stack = []
        n = len(nums)
        res = [-1 for _ in xrange(n)]
        for i in xrange(2*n):
            num = nums[i%n]
            while len(stack)!=0 and nums[stack[-1]] < num:
                res[stack.pop()] = num
            if i < n:
                stack.append(i)
                
        return res
